jobs/jobs/utils.py

```diff
@@ -581,9 +584,137 @@ async def get_job_progress(
+        return response
+
+    logger.warning(
+        f"Unknown job type: {job.type}, treating as manual by default."
+    )
+    await update_manual_job_status(session, job, response)
+    return response
+
+
+async def update_job_status_by_type(
+    session: Session,
+    job: CombinedJob,
+    response: Dict[str, int],
+    job_id: int,
+    api_client: client.ApiClient,
+) -> None:
+    job_type = job.type
+
+    if job_type in (
+        JobType.AnnotationJob,
+        JobType.ExtractionWithAnnotationJob
+    ):
+        await update_manual_job_status(session, job, response)
+    
+    else: # JobType.ExtractionJob
+        await handle_pipeline_driven_job(
+            session, job, response, job_id, api_client
+        )
+
+
+async def update_manual_job_status(
+    session: Session, job: CombinedJob, response: Dict[str, int]
+) -> None:
+    """Update the job status based on the number of completed tasks in the response.
+
+    If 'finished' > 0:
+    - Set status to 'finished' if all tasks are done.
+    - Set status to 'in_progress' if some tasks are done.
+    Otherwise, the job status remains unchanged.
+    """
+    finished = response.get("finished")
+    total = response.get("total")
+
+    if finished is None or total is None:
+        logger.warning("Missing keys in response for job %s", job.id)
+        return
+
+    if finished == 0: # If no tasks are completed, the job will remain in a pending state.
+        return
+
+    if finished == total:
+        db_service.update_job_status(session, job, Status.finished)
+    else:
+        db_service.update_job_status(session, job, Status.in_progress)
+
+
+async def handle_pipeline_driven_job(
+    session: Session,
+    job: CombinedJob,
+    response: Dict[str, int],
+    job_id: int,
+    api_client: client.ApiClient,
+) -> None:
+    """Handle Extraction jobs that depend on pipeline execution."""
+    pipeline_id = job.pipeline_id
+    dag_run_id = f"{pipeline_id}:{job_id}"
+    # To display progress for extraction jobs, response["total"] will always be 1. The value of response["finished"] will be either 0 or 1, indicating whether the pipeline has completed successfully.
+    response["total"] = 1
+
+    # Activate the pipeline if it’s not already active, this will ensure the pipeline is running.
+    await activate_pipeline(pipeline_id, api_client)
+
+    pipeline_status = await fetch_pipeline_status(pipeline_id, dag_run_id, api_client)
+
+    new_state = Status.pending
+    pipeline_completion = 0
+    if pipeline_status == AirflowPipelineStatus.success.value:
+        new_state = Status.finished
+        pipeline_completion = 1
+    elif pipeline_status == AirflowPipelineStatus.failed.value:
+        new_state = Status.failed
+        pipeline_completion = 0
+    elif pipeline_status == AirflowPipelineStatus.running.value:
+        new_state = Status.in_progress
+        pipeline_completion = 0
```

Comment:
you already set pipeline_completion = 0 outside of if statements, no need to assign it here again